### ISSUE TITLE
VSCode instructions did not work

### ISSUE THREAD

**Issue by wagenet** [gitKrystan/prettier-plugin-ember-template-tag#38]
To get things to work with VSCode I had to make the following changes:

```jsonc
// .vscode/settings.json
{
  "[glimmer-js]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode",
    "editor.formatOnSave": true
  },
  "[glimmer-ts]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode",
    "editor.formatOnSave": true
  },
  "prettier.documentSelectors": ["**/*.gjs", "**/*.gts"]
}
```
Note that these are glimmer-js and glimmer-ts not gjs and gts.

```js
// prettierrc.js
module.exports = {
	plugins: ['prettier-plugin-ember-template-tag'],
       overrides: [
		{
			files: '*.gjs',
			options: {
				printWidth: 80,
				singleQuote: false,
				parser: 'ember-template-tag',
			},
		},
		{
			files: '*.gts',
			options: {
				printWidth: 80,
				singleQuote: false,
				parser: 'ember-template-tag',
			},
		},
       }
};
```

**Comment by gitKrystan** [comment]
The `overrides` section of prettierrc.js was also necessary for getting things working in VSCode? I'm pretty surprised to see that since the plugin should already register the parser for those file types.

**Comment by wagenet** [comment]
@gitKrystan Yes, I needed all of it. It was surprising to me. There may be something else going on, but I'm not sure what.

**Comment by gitKrystan** [comment]
What glimmer-related VSCode extensions do you have installed?

**Comment by simonihmig** [comment]
I can confirm this! Struggled for a while to get it working. When explicitly calling `Format Document` in Code I was getting this:

<img width="355" alt="image" src="https://user-images.githubusercontent.com/1325249/212759495-458e30e1-9617-411b-94dc-2ce131f3a599.png">

With all of the config added that @wagenet posted above, it is finally working! :tada:

> What glimmer-related VSCode extensions do you have installed?

The Ember/Glimmer specific ones I have are
* `lifeart.vscode-ember-unstable`
* `chiragpat.vscode-glimmer`
* `typed-ember.glint-vscode`

**Comment by Techn1x** [comment]
We encountered the above only after switching from yarn 1 to pnpm 8...... So I am wondering if the OP and Simon are also using pnpm. That might be the key here, or a giant red herring.

Specifically, the config I needed was;
* the parser overrides in prettierrc.js
* and the `prettier.documentSelectors` in vscode settings

**Comment by Techn1x** [comment]
Might be stating the obvious, but without the above, when I do a `Format Document (forced)` in vscode, the `inferredParser` value is `null`
![image](https://github.com/gitKrystan/prettier-plugin-ember-template-tag/assets/1049837/fca18666-22e2-4337-ba37-2330edd80fc0)

**Comment by Techn1x** [comment]
I run vscode over a docker container (via the remote-containers extension) so there's local extensions and remote extensions, but here you go. Also added versions.

```
$ code --list-extensions --show-versions
chiragpat.vscode-glimmer@0.0.6
csstools.postcss@1.0.9
mjmcloug.vscode-elixir@1.1.0
ms-vscode-remote.remote-containers@0.292.0
vayan.haml@1.0.1

$ code --list-extensions --show-versions
Extensions installed on Container bx-docker:latest (bx-docker):
bradlc.vscode-tailwindcss@0.9.11
ChakrounAnas.turbo-console-log@2.9.3
cssho.vscode-svgviewer@2.0.0
dbaeumer.vscode-eslint@2.4.0
eamodio.gitlens@13.6.0
EditorConfig.EditorConfig@0.16.4
esbenp.prettier-vscode@9.13.0
lifeart.vscode-ember-unstable@3.0.51
mjmcloug.vscode-elixir@1.1.0
redhat.vscode-yaml@1.13.0
stylelint.vscode-stylelint@1.2.4
typed-ember.glint-vscode@1.0.3
```

**Comment by gitKrystan** [comment]
Are you all using PNPM (and are you using PNPM workspaces)?

I found some issues around how Prettier auto-loads plugins that seem like they would conflict with PNPM. Also found this in the changelog: https://github.com/prettier/prettier/blob/fc6419ab63a8b3bc2b4458bb5771fe64638ab013/changelog_unreleased/api/14759.md?plain=1#L3

**Comment by Techn1x** [comment]
Pnpm yes, workspaces no.

I did actually encounter that prettier/pnpm/plugin searching as a separate issue (after switching to pnpm, prettier couldn't find my tailwindcss prettier plugin) but listing it explicitly in prettierrc plugins resolved that.

**Comment by gitKrystan** [comment]
Notes from my investigation today:

`getSupportInfo()` (which is what prettier-vscode uses to load the supported file extensions) does not return file extensions for custom registered languages
https://github.com/prettier/prettier/issues/10561

> Plugins are automatically loaded if you have them installed in the same node_modules directory where prettier is located. Plugin package names must start with @prettier/plugin- or prettier-plugin- or @<scope>/prettier-plugin- to be registered.
-- https://prettier.io/docs/en/plugins.html#using-plugins

^^ In the case of my accidental repro (which is in a project using pnpm workspaces), the plugin is not in the same node modules folder

More plugin autoload issues:
https://github.com/prettier/prettier/issues/8474
https://github.com/prettier/prettier-vscode/issues/2586

Other plugin loading issues:
https://github.com/prettier/prettier/issues/13276

Some people had luck with something like:
```
module.exports = {
  plugins: [require('prettier-plugin-ember-template-tag')]
}
```
but I haven't gotten that working in my repro yet.

**Comment by NullVoxPopuli** [comment]
That reminds me!,
For pnpm users, this is important information: https://pnpm.io/npmrc#public-hoist-pattern

If fou only run prettier from one directory, say a monorepo root, to make the prettier plugin work like all the other prettier plugins, you need to add the plugin to the public hoist pattern, if it's not also installed at the root of the monorepo.

The issue doesn't present itself if you run prettier from each directory, as prettier and the plugin are in the same node modules (local to the package)

**Comment by Techn1x** [comment]
I think that issue you linked https://github.com/prettier/prettier/issues/13276 is exactly what the problem is. The workaround in that issue for VSCode are even the same as the OP gave here. The only difference is they use Yarn PnP rather than pnpm but it's kind of the same thing in that node_modules is a non-standard structure

> Some people had luck with something like:
> module.exports = {
>  plugins: [require('prettier-plugin-ember-template-tag')]
> }
> but I haven't gotten that working in my repro yet.

Note that providing a string to the prettier plugins array is essentially the same as calling require

https://prettier.io/docs/en/plugins.html#using-plugins
> Strings provided to plugins are ultimately passed to require(), so you can provide a module/package name, a path, or anything else require() takes.

**Comment by gitKrystan** [comment]
Can y'all try out v1.1.0 and see if any of your issues are resolved?

Specifically, in @wagenet's config I was able to remove:
`"prettier.documentSelectors": ["**/*.gjs", "**/*.gts"]` from `.vscode/settings.json` and the entire overrides section from `prettierrc.js`.

**Comment by Techn1x** [comment]
It looks like I had already removed documentSelectors from my settings (I recently changed systems and had to redo all this stuff).

I've now removed the ember-template-tag parser overrides section from my prettierrc.js

Looks like vscode prettier format on save works!

FWIW, the relevant vscode settings I have now are this.
```json
  "[glimmer-js]": {
    "editor.formatOnSave": true,
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
  "[glimmer-ts]": {
    "editor.formatOnSave": true,
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
```

**Comment by amk221** [comment]
In case this helps anybody...

Our situation is 

```
parent
  ├─ child
  │    ├─ .prettierrc.js // ember project
  ├─ .prettierrc.js // non-ember project

```

If you are working inside the `child` directory, which is an Ember app with associated prettier config with template-tag support, than it works as per the readme instructions on this repo.

But, if you are on a file inside the `child` directory, but from the `parent` directory in VSCode, then Format on Save does not work.

To fix it I needed to add

```
"prettier.documentSelectors": ["**/*.gjs"]
```

I did not need any of the previously mentioned `[glimmer-js]` configs.

**Comment by jaswilli** [comment]
Hey all, just adding another report:

I'm using plugin version `1.1.0` and had to add:

`"prettier.documentSelectors": ["**/*.gjs", "**/*.gts"]`

I already had the below settings:
```
"[glimmer-js]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode",
    "editor.foldingStrategy": "indentation"
  },
  "[glimmer-ts]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode",
    "editor.foldingStrategy": "indentation"
  },
},
```

<details>
  <summary>In case it's useful, here's my `code --list-extensions`</summary>
  <pre>
amazonwebservices.aws-toolkit-vscode
attilabuti.vscode-mjml
bierner.markdown-preview-github-styles
bradlc.vscode-tailwindcss
castwide.solargraph
dbaeumer.vscode-eslint
denoland.vscode-deno
eamodio.gitlens
EditorConfig.EditorConfig
Equinusocio.vsc-material-theme
equinusocio.vsc-material-theme-icons
esbenp.prettier-vscode
GitHub.copilot
GitHub.copilot-chat
golang.go
JakeBecker.elixir-ls
jprestidge.theme-material-theme
jtr.vscode-position
kaiwood.endwise
lifeart.vscode-ember-unstable
lifeart.vscode-glimmer-syntax
ms-azuretools.vscode-docker
ms-python.black-formatter
ms-python.flake8
ms-python.isort
ms-python.mypy-type-checker
ms-python.python
ms-python.vscode-pylance
ms-toolsai.jupyter
ms-toolsai.jupyter-keymap
ms-toolsai.jupyter-renderers
ms-toolsai.vscode-jupyter-cell-tags
ms-toolsai.vscode-jupyter-slideshow
ms-vscode-remote.remote-containers
ms-vscode.makefile-tools
ms-vscode.sublime-keybindings
msjsdiag.vscode-react-native
octref.vetur
PKief.material-icon-theme
redhat.vscode-commons
redhat.vscode-yaml
rubocop.vscode-rubocop
Rubymaniac.vscode-direnv
rust-lang.rust-analyzer
samuelcolvin.jinjahtml
Shopify.ruby-lsp
stylelint.vscode-stylelint
svelte.svelte-vscode
tamasfe.even-better-toml
timonwong.shellcheck
typed-ember.glint-vscode
vadimcn.vscode-lldb
VisualStudioExptTeam.intellicode-api-usage-examples
VisualStudioExptTeam.vscodeintellicode
<pre>
</details>
